src/extendedstorage.cpp

```diff
@@ -142,25 +158,27 @@ void ExtendedStorage::clearLoaded()
 }
 
 bool ExtendedStorage::getLoadDates(const QDate &start, const QDate &end,
-                                   QDateTime &loadStart, QDateTime &loadEnd)
+                                   QDateTime &loadStart, QDateTime &loadEnd) const
 {
-    // Check the need to load from db.
-    if (start.isValid() && d->mStart.isValid() && start >= d->mStart &&
-            end.isValid() && d->mEnd.isValid() && end <= d->mEnd) {
-        return false;
-    }
+    loadStart.setDate(start);   // may be null if start is not valid
+    loadEnd.setDate(end);   // may be null if end is not valid
 
-    // Set load dates to load only what's necessary.
-    if (start.isValid() && d->mStart.isValid() && start >= d->mStart) {
-        loadStart.setDate(d->mEnd);
-    } else {
-        loadStart.setDate(start);   // may be null if start is not valid
+    // Check the need to load from db.
+    for (const Range &loadedRange : d->mRanges) {
+        bool startIsIn = loadedRange.contains(loadStart.date())
+            || (loadedRange.mStart.isNull() && loadStart.date().isNull());
+        bool endIsIn = loadedRange.contains(loadEnd.date())
+            || (loadedRange.mEnd.isNull() && loadEnd.date().isNull());
+        if (startIsIn && endIsIn) {
+            return false;
+        } else if (startIsIn) {
+            loadStart.setDate(loadedRange.mEnd);
```

Comment:
That's a good point. Not completely sure here myself. As arguments, start is inclusive and end is exclusive (`load(QDate a) <=> load(QDate a, QDate a+1)`).

So at first, I simply put `at >= mStart && at < mEnd` in the Range::contains() routine. But with QDates, this is not working for instance with a saved range R = [5, 9[ and asking to load [1, 9[, it will not result with loading [1, 5[ (because 9 is not seen as contained by R). So I put a contains test on [x, y] and not [x, y[.

This is not creating an immediate issue since loadStart and loadEnd arguments are QDateTime, with QTime being 00:00. So having QDate a <= x < QDate b is equivalent to QDateTime(a, {}) <= x <= QDatetime(b, {}).addSecs(-1).

So at one second rounded on the end part, we're doing the exact comparison here. I could remove one second on the end part after the for loop. That could be a good idea. Or I could change the SQL request in sqlitestorage.h to do strict comparison for end parts… Not sure it actually matters much. On the worst case, in a load [a, b[ we're now loading events starting at midnight the day b, while call load(b, c) will reload this midnight event at b.

What do you prefer ?

Reply:
> As arguments, start is inclusive and end is exclusive 

Actually that looks like not being entirely defined. ExtendedStorage::load(start,end) says "Load incidences between given dates into the memory.", not saying what between means exactly. But then ::load(const QDate &date) does seem to use it as in latter parameter being exclusive. And one case in nemo-calendar now. If it helps suppose the definition can be still changed but at least better if method defined what the between means.

> So at first, I simply put at >= mStart && at < mEnd in the Range::contains() routine. But with QDates, this is not working for instance with a saved range R = [5, 9[ and asking to load [1, 9[, it will not result with loading [1, 5[ (because 9 is not seen as contained by R). So I put a contains test on [x, y] and not [x, y[.

Alright, the Range doesn't need to match the definition of the method if it helps. But good to note if differences. If these are defined different, suppose this getLoadDates() should then take one day away when comparing to ranges.

> What do you prefer ?

Guess most of all I'm preferring defined meanings on the end date inclusivity. Going through the code I get a feeling it might be easier to accomplish by adjusting the sql a little, but at this moment emphasis on the feeling part.